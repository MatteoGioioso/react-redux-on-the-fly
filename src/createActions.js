/**
 * createActions
 * @param {String} section
 * @param {Object} options you can inject additional types and action
 * Create a template with common actions and types. Just need to pass a section name
 * and register it into asyncReducersActions.js file
 */
export const createActions = (
  section,
  options = { types: {}, actions: {} }
) => ({
  types: {
    FIND: `FIND_${section}S`,
    FIND_BY_ID: `FIND_${section}_BY_ID`,
    CREATE: `CREATE_${section}`,
    UPDATE_BY_ID: `UPDATE_${section}_BY_ID`,
    DELETE_BY_ID: `DELETE_${section}_BY_ID`,
    RECEIVE_MANY: `RECEIVE_${section}S`,
    RECEIVE_MANY_ON_TOP: `RECEIVE_${section}S_ON_TOP`,
    RECEIVE_ONE: `RECEIVE_${section}`,
    SUBSTITUTE_ONE: `SUBSTITUTE_${section}`,
    REMOVE_ONE: `REMOVE_${section}`,
    RESET: `RESET_${section}S`,
    ...options.types
  },

  // The only way to reference the same Object's properties during its creation is to use a getter
  get actions() {
    return {
      [`find${this._normalizeName()}s`]: (urlObject, callback) => ({
        type: this.types.FIND,
        payload: { urlObject, callback }
      }),
      [`find${this._normalizeName()}ById`]: (urlObject, callback) => ({
        type: this.types.FIND_BY_ID,
        payload: { urlObject, callback }
      }),
      [`create${this._normalizeName()}`]: (urlObject, data, callback) => ({
        type: this.types.CREATE,
        payload: { urlObject, data, callback }
      }),
      [`update${this._normalizeName()}ById`]: (urlObject, data, callback) => ({
        type: this.types.UPDATE_BY_ID,
        payload: { urlObject, data, callback }
      }),
      [`delete${this._normalizeName()}ById`]: (urlObject, data, callback) => ({
        type: this.types.DELETE_BY_ID,
        payload: { urlObject, data, callback }
      }),

      [`receive${this._normalizeName()}s`]: (data, name) => ({
        type: this.types.RECEIVE_MANY,
        name,
        payload: data
      }),
      [`receive${this._normalizeName()}sOnTop`]: (data, name) => ({
        type: this.types.RECEIVE_MANY_ON_TOP,
        name,
        payload: data
      }),
      [`substitute${this._normalizeName()}`]: (data, name) => ({
        type: this.types.SUBSTITUTE_ONE,
        name,
        payload: data
      }),
      [`remove${this._normalizeName()}`]: (id, name) => ({
        type: this.types.REMOVE_ONE,
        name,
        payload: id
      }),

      [`receive${this._normalizeName()}`]: (data, name) => ({
        type: this.types.RECEIVE_ONE,
        name,
        payload: data
      }),

      [`reset${this._normalizeName()}s`]: name => ({
        type: this.types.RESET,
        name
      }),

      ...options.actions
    };
  },

  /**
   * _normalizeName
   * @returns {string}
   * @private
   * input "EVENT_AVAILABLE" -> output "EventAvailable"
   */
  _normalizeName: () => {
    const capitalizeFirstLetter = name => {
      const firstCapitalLetter = name[0];
      const restOfTheNameLowerCase = name.slice(1).toLowerCase();

      return firstCapitalLetter + restOfTheNameLowerCase;
    };

    if (section.includes("_")) {
      return section
        .split("_")
        .map(name => capitalizeFirstLetter(name))
        .join("");
    } else {
      return capitalizeFirstLetter(section);
    }
  }
});
