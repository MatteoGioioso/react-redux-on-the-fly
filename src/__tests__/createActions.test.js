import { createActions } from "../createActions";

describe("actionsCreator", () => {
  it("should return types and actions correctly", function() {
    const result = createActions("CUSTOMER");

    expect(result.actions).toMatchSnapshot();
    expect(result.types).toMatchSnapshot();
  });

  it("should add more types and actions", function() {
    const options = {
      types: {
        CUSTOM_TYPE: "CUSTOM_TYPE"
      },
      actions: {
        myCustomAction: data => ({ type: "CUSTOM_TYPE", payload: data })
      }
    };
    const result = createActions("CUSTOMER", options);

    expect(result.actions).toMatchSnapshot();
    expect(result.types).toMatchSnapshot();
    expect(result.actions.myCustomAction).toBeDefined();
  });
});
