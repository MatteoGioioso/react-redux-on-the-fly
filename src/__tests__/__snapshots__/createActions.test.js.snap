// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actionsCreator should add more types and actions 1`] = `
Object {
  "createCustomer": [Function],
  "deleteCustomerById": [Function],
  "destroyCustomer": [Function],
  "destroyCustomers": [Function],
  "findCustomerById": [Function],
  "findCustomers": [Function],
  "myCustomAction": [Function],
  "receiveCustomer": [Function],
  "receiveCustomerOnTop": [Function],
  "receiveCustomers": [Function],
  "receiveCustomersOnTop": [Function],
  "resetCustomers": [Function],
  "substituteCustomer": [Function],
  "updateCustomerById": [Function],
}
`;

exports[`actionsCreator should add more types and actions 2`] = `
Object {
  "CREATE": "CREATE_CUSTOMER",
  "CUSTOM_TYPE": "CUSTOM_TYPE",
  "DELETE_BY_ID": "DELETE_CUSTOMER_BY_ID",
  "DESTROY_MANY": "DESTROY_CUSTOMERS",
  "DESTROY_ONE": "DESTROY_CUSTOMER",
  "FIND": "FIND_CUSTOMERS",
  "FIND_BY_ID": "FIND_CUSTOMER_BY_ID",
  "RECEIVE_MANY": "RECEIVE_CUSTOMERS",
  "RECEIVE_MANY_ON_TOP": "RECEIVE_CUSTOMERS_ON_TOP",
  "RECEIVE_ONE": "RECEIVE_CUSTOMER",
  "RECEIVE_ONE_ON_TOP": "RECEIVE_CUSTOMER_ON_TOP",
  "RESET": "RESET_CUSTOMERS",
  "SUBSTITUTE_ONE": "SUBSTITUTE_CUSTOMER",
  "UPDATE_BY_ID": "UPDATE_CUSTOMER_BY_ID",
}
`;

exports[`actionsCreator should return types and actions correctly 1`] = `
Object {
  "createCustomer": [Function],
  "deleteCustomerById": [Function],
  "destroyCustomer": [Function],
  "destroyCustomers": [Function],
  "findCustomerById": [Function],
  "findCustomers": [Function],
  "receiveCustomer": [Function],
  "receiveCustomerOnTop": [Function],
  "receiveCustomers": [Function],
  "receiveCustomersOnTop": [Function],
  "resetCustomers": [Function],
  "substituteCustomer": [Function],
  "updateCustomerById": [Function],
}
`;

exports[`actionsCreator should return types and actions correctly 2`] = `
Object {
  "CREATE": "CREATE_CUSTOMER",
  "DELETE_BY_ID": "DELETE_CUSTOMER_BY_ID",
  "DESTROY_MANY": "DESTROY_CUSTOMERS",
  "DESTROY_ONE": "DESTROY_CUSTOMER",
  "FIND": "FIND_CUSTOMERS",
  "FIND_BY_ID": "FIND_CUSTOMER_BY_ID",
  "RECEIVE_MANY": "RECEIVE_CUSTOMERS",
  "RECEIVE_MANY_ON_TOP": "RECEIVE_CUSTOMERS_ON_TOP",
  "RECEIVE_ONE": "RECEIVE_CUSTOMER",
  "RECEIVE_ONE_ON_TOP": "RECEIVE_CUSTOMER_ON_TOP",
  "RESET": "RESET_CUSTOMERS",
  "SUBSTITUTE_ONE": "SUBSTITUTE_CUSTOMER",
  "UPDATE_BY_ID": "UPDATE_CUSTOMER_BY_ID",
}
`;
